### PR TITLE
Indicate elided text with "…" & related formatting tweaks

### LINKED ISSUE #637: format string with two levels of conditions and long format key does not allow width specifier
`cmus --version`:
cmus v2.8.0-rc0

It’s me again with another string format issue. :) I found that I cannot specify a width for an item if it is
- within two levels of conditions and
- a long keyword, i.e. one that uses braces

I.o.w., the following format strings work:
`:set format_trackwin=%{?date?%{?date?%{date}}}`
`:set format_trackwin=%{?date?%{?date?%4y}}`

But this does not, even though it should be identical to the last one mentioned above:
`:set format_trackwin=%{?date?%{?date?%4{date}}}`

### LINKED ISSUE #1004: String formatting lacks support for minimum width.
The man page for cmus says:

> You can use printf style formatting (width, alignment, padding).

However, when I try to use `%.30{album}` as a part of the format string for "format_current" (to show 30 characters or less of the current album name), is tells me this is an invalid format string.

1) `uname -a`:
```
Linux Fillip 5.4.0-54-generic #60-Ubuntu SMP Fri Nov 6 10:37:59 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

2) `cmus --version`:
```
cmus v2.8.0
Copyright 2004-2006 Timo Hirvonen
Copyright 2008-2016 Various Authors)
```

3) `cmus --plugins`:
```
Input Plugins: /usr/lib/cmus/ip
  mad:
    Priority: 55
    File Types: mp3 mp2
    MIME Types: audio/mpeg audio/x-mp3 audio/x-mpeg
  cdio:
    Priority: 50
    File Types:
    MIME Types: x-content/audio-cdda
  cue:
    Priority: 50
    File Types:
    MIME Types: application/x-cue
  opus:
    Priority: 50
    File Types: opus
    MIME Types:
  aac:
    Priority: 50
    File Types: aac
    MIME Types: audio/aac audio/aacp
  mpc:
    Priority: 50
    File Types: mpc mpp mp+
    MIME Types: audio/x-musepack
  wavpack:
    Priority: 50
    File Types: wv
    MIME Types: audio/x-wavpack
  vorbis:
    Priority: 50
    File Types: ogg oga ogx
    MIME Types: application/ogg audio/x-ogg
  modplug:
    Priority: 50
    File Types: mod s3m xm it 669 amf ams dbm dmf dsm far mdl med mtm okt ptm stm ult umx mt2 psm
    MIME Types:
  flac:
    Priority: 50
    File Types: flac fla
    MIME Types:
  wav:
    Priority: 50
    File Types: wav
    MIME Types:
  ffmpeg:
    Priority: 30
    File Types: aa aac ac3 aif aifc aiff ape au fla flac m4a m4b mka mkv mp+ mp2 mp3 mp4 mpc mpp ogg shn tak tta wav webm wma wv
    MIME Types:

Output Plugins: /usr/lib/cmus/op
  pulse
  alsa
  jack
  ao
```

### PR DESCRIPTION
**Pull request by gavtroy**
Closes #637.
Closes #1004.

### CODE REVIEW

**Review by pgaskin — COMMENTED** [review]
Looks good, but not tested yet.

File: options.c [hunk 1]
```diff
@@ -1602,6 +1619,15 @@ void options_load(void)
...
 	}
+
+	/* check if charset supports the default format_clipped_text symbol */
+	if (strcmp(clipped_text_format, str_defaults[FMT_CLIPPED_TEXT].value) == 0) {
+		mbstate_t mbstate = {0};
+		char mb[MB_CUR_MAX];
+		if (wcrtomb(mb, L'…', &mbstate) == -1) {
+			clipped_text_internal = xstrdup("...");
+		}
+	}
```
Comment (pgaskin) [inline options.c:1630]: I haven't looked into the cause yet, but I noticed that left playlist pane and the file browser always seem to use `...` for some reason, even when the proper ellipsis is supported in other areas.

Comment (gavtroy) [reply options.c:1630]: Good catch. These places use format_str() from ui_curses.c which already adds "..." even in the current version.

**Comment by gavtroy** [comment]
Sorry to add so much churn at this stage. There are so many minor details to formatting. Hopefully I've got everything now.

One thing I notice but haven't bothered fixing is that we use a hardcoded 512 byte buffer, but format_str simply assumes there'll always be space. Cmus will crash if your terminal is wider than this. 512 isn't a particularly large limit, considering on my 2560x1440 monitor with a font width of 8 I already reach 320 columns, and ultrawide monitors exist. Each multibyte character in the Browser (if open) will also reduce the threshold needed to cause a crash, although for most users multibyte characters in paths are going to be rare.
Since this is not a new problem with my changes I think I will just leave it alone for now.

**Comment by pgaskin** [comment]
The new changes appear to work well so far and the code makes sense based on a quick glance, but I'll need to do more testing over the next few days.

> Sorry to add so much churn at this stage

It's OK (but I really should see if there's a better way to merge force-pushed branches without having to manually ensure there aren't leftovers from the previous merge).

> we use a hardcoded 512 byte buffer, but format_str simply assumes there'll always be space

Yes, there's a bunch of those left behind from before. It'd probably be a good idea to deal with that properly in a separate PR.

> Each multibyte character in the Browser

Unrelated to this PR, but speaking of multibyte characters, I need to figure out why combining characters can cause the line to wrap over to the next one.

**Review by pgaskin — APPROVED** [review]
LGTM

**Comment by pgaskin** [comment]
@nefthy?

**Comment by pgaskin** [comment]
@flyingmutant

**Comment by flyingmutant** [comment]
Do you consider this ready to merge?

**Comment by pgaskin** [comment]
Yes, the code (including the width calculations) looks fine, it works fine with a bunch of characters I tested, it works consistently across all views, and I've been using it for the last month.

**Comment by flyingmutant** [comment]
Merged, thanks a lot!

**Comment by gavtroy** [comment]
Thanks. It's now squashed into a single commit; for reference I have been aiming for logically separated commits in my pull requests, though I realize this probably isn't the typical GitHub workflow.

I've followed up with #1065 for the mentioned pre-existing issue with very wide windows.

**Comment by flyingmutant** [comment]
Yep, sorry about the squash -- more often than not people are not very careful about separating the commits, so I've made a (questionable) habit of squashing to keep history understandable. But, on the positive side, it helps to make git shortlog closer to the release notes.

**Pull request merged**